File: data/analyzeTest.js
const fs = require('fs/promises');
const path = require('path');

const JSON_DIR = './tests';

async function collectData() {
    try {
        const files = await fs.readdir(JSON_DIR);
        const jsonFiles = files.filter(file => file.endsWith('.json'));

        // Initialize accumulators for all metrics
        const metrics = {
            length: { total: 0, sum_squares: 0 },
            deltaX: { total: 0, sum_squares: 0 },
            deltaY: { total: 0, sum_squares: 0 },
            deltaAngleDegrees: { total: 0, sum_squares: 0 },
            speed_entropy: { total: 0, sum_squares: 0 },
            acceleration_entropy: { total: 0, sum_squares: 0 },
            braking_entropy: { total: 0, sum_squares: 0 },
            positions_mean: { total: 0, sum_squares: 0 },
            avg_radius_mean: { total: 0, sum_squares: 0 },
            gaps_mean: { total: 0, sum_squares: 0 },
            right_bends: { total: 0, sum_squares: 0 },
            avg_radius_var: { total: 0, sum_squares: 0 },
            total_overtakes: { total: 0, sum_squares: 0 },
            straight_sections: { total: 0, sum_squares: 0 },
            gaps_var: { total: 0, sum_squares: 0 },
            left_bends: { total: 0, sum_squares: 0 },
            positions_var: { total: 0, sum_squares: 0 },
            curvature_entropy: { total: 0, sum_squares: 0 }
        };

        let count = 0;
        let perfectTrackCount = 0;
        let veryBadTrackCount = 0;
        let perfectTracks = [];
        let veryBadTracks = [];

        // First pass: collect sums for averages
        for (const file of jsonFiles) {
            const filePath = path.join(JSON_DIR, file);
            const data = JSON.parse(await fs.readFile(filePath, 'utf-8'));
            
            if (!data.fitness) continue; // Skip if no fitness data

            const fitness = data.fitness;
            count++;

            // Update all metrics
            for (const [key, value] of Object.entries(metrics)) {
                if (fitness[key] !== null && fitness[key] !== undefined) {
                    value.total += fitness[key];
                }
            }

            // Track quality checks
            if (Math.abs(fitness.deltaX) < 1 && 
                Math.abs(fitness.deltaY) < 1 && 
                Math.abs(fitness.deltaAngleDegrees) < 10) {
                perfectTrackCount++;
                perfectTracks.push({
                    seed: data.id,
                    trackSize: data.trackSize,
                    mode: data.mode,
                    metrics: { ...fitness }
                });
            }

            if ((Math.abs(fitness.deltaX) + Math.abs(fitness.deltaY) > 6)) {
                veryBadTrackCount++;
                veryBadTracks.push({
                    seed: data.id,
                    trackSize: data.trackSize,
                    mode: data.mode,
                    metrics: { ...fitness }
                });
            }
        }

        // Calculate averages
        const averages = {};
        for (const [key, value] of Object.entries(metrics)) {
            averages[key] = value.total / count;
        }

        // Second pass: calculate variances
        for (const file of jsonFiles) {
            const filePath = path.join(JSON_DIR, file);
            const data = JSON.parse(await fs.readFile(filePath, 'utf-8'));
            
            if (!data.fitness) continue;
            
            const fitness = data.fitness;

            // Update sum of squares for all metrics
            for (const [key, value] of Object.entries(metrics)) {
                if (fitness[key] !== null && fitness[key] !== undefined) {
                    value.sum_squares += Math.pow(fitness[key] - averages[key], 2);
                }
            }
        }

        // Calculate standard deviations
        const stdDevs = {};
        for (const [key, value] of Object.entries(metrics)) {
            stdDevs[key] = Math.sqrt(value.sum_squares / (count - 1));
        }

        // Print results
        console.log(`\nAnalysis Results (${count} valid tracks):`);
        console.log('\nBasic Statistics:');
        for (const [key, avg] of Object.entries(averages)) {
            console.log(`${key}:`);
            console.log(`  Average: ${avg.toFixed(4)}`);
            console.log(`  Std Dev: ${stdDevs[key].toFixed(4)}`);
        }

        console.log('\nTrack Quality Analysis:');
        console.log(`Perfect Tracks: ${perfectTrackCount}`);
        console.log(`Problematic Tracks: ${veryBadTrackCount}`);

        if (perfectTrackCount > 0) {
            console.log('\nPerfect Tracks Details:');
            perfectTracks.forEach(track => {
                console.log(`\nSeed: ${track.seed}`);
                console.log(`Track Size: ${track.trackSize}`);
                console.log(`Mode: ${track.mode}`);
                console.log('Key Metrics:');
                console.log(`  Length: ${track.metrics.length.toFixed(2)}`);
                console.log(`  Speed Entropy: ${track.metrics.speed_entropy.toFixed(2)}`);
                console.log(`  Curvature Entropy: ${track.metrics.curvature_entropy.toFixed(2)}`);
            });
        }

        if (veryBadTrackCount > 0) {
            console.log('\nProblematic Tracks Details:');
            veryBadTracks.forEach(track => {
                console.log(`\nSeed: ${track.seed}`);
                console.log(`Track Size: ${track.trackSize}`);
                console.log(`Mode: ${track.mode}`);
                console.log('Issues:');
                console.log(`  Delta X: ${track.metrics.deltaX.toFixed(2)}`);
                console.log(`  Delta Y: ${track.metrics.deltaY.toFixed(2)}`);
                console.log(`  Delta Angle: ${track.metrics.deltaAngleDegrees.toFixed(2)}`);
            });
        }

    } catch (err) {
        console.error(`Error: ${err.message}`);
    }
}

collectData();

File: src/data/analyzeTest.py
import os
import json
import math

JSON_DIR = './tests'

def collect_data():
    try:
        files = os.listdir(JSON_DIR)
        json_files = [f for f in files if f.endswith('.json')]

        # Initialize accumulators for all metrics
        metrics = {
            'length': {'total': 0, 'sum_squares': 0},
            'deltaX': {'total': 0, 'sum_squares': 0},
            'deltaY': {'total': 0, 'sum_squares': 0},
            'deltaAngleDegrees': {'total': 0, 'sum_squares': 0},
            'speed_entropy': {'total': 0, 'sum_squares': 0},
            'acceleration_entropy': {'total': 0, 'sum_squares': 0},
            'braking_entropy': {'total': 0, 'sum_squares': 0},
            'positions_mean': {'total': 0, 'sum_squares': 0},
            'avg_radius_mean': {'total': 0, 'sum_squares': 0},
            'gaps_mean': {'total': 0, 'sum_squares': 0},
            'right_bends': {'total': 0, 'sum_squares': 0},
            'avg_radius_var': {'total': 0, 'sum_squares': 0},
            'total_overtakes': {'total': 0, 'sum_squares': 0},
            'straight_sections': {'total': 0, 'sum_squares': 0},
            'gaps_var': {'total': 0, 'sum_squares': 0},
            'left_bends': {'total': 0, 'sum_squares': 0},
            'positions_var': {'total': 0, 'sum_squares': 0},
            'curvature_entropy': {'total': 0, 'sum_squares': 0}
        }

        count = 0
        perfect_track_count = 0
        very_bad_track_count = 0
        perfect_tracks = []
        very_bad_tracks = []

        # First pass: collect sums for averages
        for file in json_files:
            file_path = os.path.join(JSON_DIR, file)
            with open(file_path, 'r', encoding='utf-8') as f:
                data = json.load(f)
            
            # Check if data is valid and has a non-None 'fitness'
            fitness = data.get('fitness')
            if fitness is None:
                continue

            count += 1

            # Update all metrics
            for key in metrics.keys():
                if key in fitness and fitness[key] is not None:
                    metrics[key]['total'] += fitness[key]

            # Track quality checks
            if (abs(fitness.get('deltaX', 0)) < 1 and
                abs(fitness.get('deltaY', 0)) < 1 and
                abs(fitness.get('deltaAngleDegrees', 0)) < 10):
                perfect_track_count += 1
                perfect_tracks.append({
                    'seed': data.get('id'),
                    'trackSize': data.get('trackSize'),
                    'mode': data.get('mode'),
                    'metrics': fitness.copy()
                })

            if (abs(fitness.get('deltaX', 0)) + abs(fitness.get('deltaY', 0)) > 6):
                very_bad_track_count += 1
                very_bad_tracks.append({
                    'seed': data.get('id'),
                    'trackSize': data.get('trackSize'),
                    'mode': data.get('mode'),
                    'metrics': fitness.copy()
                })

        # Calculate averages
        averages = {}
        for key, value in metrics.items():
            averages[key] = value['total'] / count if count else 0

        # Second pass: calculate variances
        for file in json_files:
            file_path = os.path.join(JSON_DIR, file)
            with open(file_path, 'r', encoding='utf-8') as f:
                data = json.load(f)
            
            # Check again if data is valid and has a non-None 'fitness'
            fitness = data.get('fitness')
            if fitness is None:
                continue

            for key in metrics.keys():
                if key in fitness and fitness[key] is not None:
                    metrics[key]['sum_squares'] += (fitness[key] - averages[key]) ** 2

        # Calculate standard deviations
        std_devs = {}
        for key, value in metrics.items():
            std_devs[key] = math.sqrt(value['sum_squares'] / (count - 1)) if count > 1 else 0

        # Print results
        print(f"\nAnalysis Results ({count} valid tracks):")
        print("\nBasic Statistics:")
        for key in averages:
            print(f"{key}:")
            print(f"  Average: {averages[key]:.4f}")
            print(f"  Std Dev: {std_devs[key]:.4f}")

        print("\nTrack Quality Analysis:")
        print(f"Perfect Tracks: {perfect_track_count}")
        print(f"Problematic Tracks: {very_bad_track_count}")

        if perfect_track_count > 0:
            print("\nPerfect Tracks Details:")
            for track in perfect_tracks:
                print(f"\nSeed: {track['seed']}")
                print(f"Track Size: {track['trackSize']}")
                print(f"Mode: {track['mode']}")
                print("Key Metrics:")
                print(f"  Length: {track['metrics'].get('length', 0):.2f}")
                print(f"  Speed Entropy: {track['metrics'].get('speed_entropy', 0):.2f}")
                print(f"  Curvature Entropy: {track['metrics'].get('curvature_entropy', 0):.2f}")

        if very_bad_track_count > 0:
            print("\nProblematic Tracks Details:")
            for track in very_bad_tracks:
                print(f"\nSeed: {track['seed']}")
                print(f"Track Size: {track['trackSize']}")
                print(f"Mode: {track['mode']}")
                print("Issues:")
                print(f"  Delta X: {track['metrics'].get('deltaX', 0):.2f}")
                print(f"  Delta Y: {track['metrics'].get('deltaY', 0):.2f}")
                print(f"  Delta Angle: {track['metrics'].get('deltaAngleDegrees', 0):.2f}")

    except Exception as e:
        print(f"Error: {e}")

if __name__ == '__main__':
    collect_data()
